core/src/sumeragi/network_topology.rs

```diff
@@ -36,90 +35,111 @@ pub struct ConsensusTopology<'topology> {
+        (self.0.len().saturating_sub(1)) / 3
     }
 
     /// The required amount of votes to commit a block with this topology.
     pub fn min_votes_for_commit(&self) -> usize {
-        let len = self.ordered_peers.len();
+        let len = self.0.len();
         if len > 3 {
             self.max_faults() * 2 + 1
         } else {
             len
         }
     }
 
-    /// Index of leader among `ordered_peers`
+    /// Index of leader
     #[allow(clippy::unused_self)] // In order to be consistent with `proxy_tail_index` method
     fn leader_index(&self) -> usize {
         0
     }
 
-    /// Index of leader among `ordered_peers`
+    /// Index of proxy tail
     fn proxy_tail_index(&self) -> usize {
-        // NOTE: proxy tail is the last element from the set A so that's why it's `min_votes_for_commit - 1`
+        // NOTE: proxy tail is the last element of set A
         self.min_votes_for_commit() - 1
     }
 
     /// Filter signatures by roles in the topology.
     #[allow(clippy::comparison_chain)]
-    pub fn filter_signatures_by_roles<'a, T: 'a, I: IntoIterator<Item = &'a SignatureOf<T>>>(
+    pub fn filter_signatures_by_roles<'a, I: IntoIterator<Item = &'a BlockSignature>>(
         &self,
         roles: &[Role],
         signatures: I,
-    ) -> Vec<SignatureOf<T>> {
-        let mut public_keys = IndexSet::with_capacity(self.ordered_peers.len());
+    ) -> Vec<BlockSignature> {
+        let mut filtered = Vec::new();
+
         for role in roles {
             match (role, self.is_non_empty(), self.is_consensus_required()) {
-                (Role::Leader, Some(topology), _) => {
-                    public_keys.insert(topology.leader().public_key());
-                }
-                (Role::ProxyTail, _, Some(topology)) => {
-                    public_keys.insert(&topology.proxy_tail().public_key);
-                }
+                (Role::Leader, Some(topology), _) => filtered.push(topology.leader_index()),
+                (Role::ProxyTail, _, Some(topology)) => filtered.push(topology.proxy_tail_index()),
                 (Role::ValidatingPeer, _, Some(topology)) => {
-                    for peer in topology.validating_peers() {
-                        public_keys.insert(peer.public_key());
-                    }
+                    filtered.extend(topology.leader_index() + 1..topology.proxy_tail_index());
                 }
                 (Role::ObservingPeer, _, Some(topology)) => {
-                    for peer in topology.observing_peers() {
-                        public_keys.insert(peer.public_key());
-                    }
+                    filtered.extend(topology.proxy_tail_index() + 1..topology.0.len());
                 }
                 _ => {}
             };
         }
+
         signatures
             .into_iter()
-            .filter(|signature| public_keys.contains(signature.public_key()))
+            .filter(|signature| filtered.contains(&(signature.0 as usize)))
```

Comment:
Looks like check that signature actually match peer's public key should be checked here as well

Reply:
ugh, good point. This will take me some time to fix

Reply:
should be fixed now, but the change propagated